R/xcmsSet.R

```diff
@@ -2296,3 +2323,157 @@ setMethod("profMethod", "xcmsSet", function(object) {
+  ## profinfo, polarity.
+  xsub <- x
+  ## first of all, subset the phenoData
+  phenoData(xsub) <- droplevels(phenoData(xsub)[j,, drop=FALSE])
+  ## then the file paths
+  filepaths(xsub) <- filepaths(x)[j]
+  ## now starting to subset data:
+  ## 1) @rt$raw, @rt$corrected
+  xsub@rt$raw <- x@rt$raw[j]
+  xsub@rt$corrected <- x@rt$corrected[j]
+  ## 2) @peaks
+  keep.peaks <- x@peaks
+  rownames(keep.peaks) <- as.character(1:nrow(keep.peaks))
+  ## subsetting the peaks. Since we want to also allow reverse ordering we have to
+  ## do it a little more complicated...
+  keep.peaks <- split(data.frame(keep.peaks), f=keep.peaks[, "sample"])
+  keep.peaks <- keep.peaks[as.character(j)]
+  names(keep.peaks) <- NULL
+  keep.peaks <- as.matrix(do.call(rbind, keep.peaks))
+  ##keep.peaks <- keep.peaks[as.character(keep.peaks[, "sample"]) %in% as.character(j), ]
+  ## have to replace the sample index.
+  newsample <- numeric(nrow(keep.peaks))
+  for(idx in 1:length(j)){
+      newsample[keep.peaks[, "sample"]==j[idx]] <- idx
+  }
+  keep.peaks[, "sample"] <- newsample
+  xsub@peaks <- keep.peaks
+  rownames(xsub@peaks) <- NULL
+  ## 3) groupidx if present. subset this to indices of peaks which we will keep.
+  ##    will use the rownames of the keep.peaks matrix for that (represents the
+  ##    original index).
+  if(length(x@groupidx) > 0){
+      keep.groupidx <- lapply(x@groupidx, function(z){
+                                  newidx <- match(as.character(z), rownames(keep.peaks))
+                                  return(newidx[!is.na(newidx)])
+                                  ## that way I just return the indices as they are!
+                                  ##return(z[as.character(z) %in% rownames(keep.peaks)])
+                              })
+      xsub@groupidx <- keep.groupidx
+      ## 4) groups have to be re-calculated.
+      keep.groups <- x@groups
+      keep.groups <- keep.groups[, -(which(colnames(keep.groups)=="npeaks"):ncol(keep.groups))]
+      sampclasses <- sampclass(xsub)
+      peakCounts <- matrix(ncol=length(levels(sampclasses)), nrow=nrow(keep.groups), 0)
+      colnames(peakCounts) <- levels(sampclasses)
+      ## loop throught the peakcounts
+      for(idx in 1:nrow(keep.groups)){
+          groupidx <- keep.groupidx[[idx]]
+          if(length(groupidx) == 0)
+              next
+          tab <- keep.peaks[groupidx, "sample"]
+          tab <- table(sampclasses[tab])
+          peakCounts[idx, names(tab)] <- as.numeric(tab)
+      }
+      keep.groups <- cbind(keep.groups, npeaks=rowSums(peakCounts), peakCounts)
+      xsub@groups <- keep.groups
+  }
+  ## 5) filled
+  if(length(x@filled) > 0){
+      xsub@filled <- (1:nrow(keep.peaks))[rownames(keep.peaks) %in% as.character(x@filled)]
+  }
+  ## 6) dataCorrection
+  if(length(x@dataCorrection)>0)
+      xset@dataCorrection <- x@dataCorrection[j]
+  return(xsub)
+})
+## setMethod("[", "eSet", function(x, i, j, ..., drop = FALSE) {
```

Comment:
a yes, right. The commented out code can be removed straight away... sorry, forgot about that.